### PR TITLE
Adds an option to normalize gradients before updating weights.

### PR DESCRIPTION
**Pull request by raghakot**
This paper: https://arxiv.org/pdf/1707.04822.pdf, shows that gradient normalization can improve convergence by alleviating vanishing/exploding gradients.

### CODE REVIEW

File: tests/keras/optimizers_test.py [hunk 1]
```diff
@@ -108,4 +113,5 @@ def test_tfoptimizer():
 
 
 if __name__ == '__main__':
-    pytest.main([__file__])
+    test_norm_grads()
+    # pytest.main([__file__])
```
Comment (raghakot) [inline tests/keras/optimizers_test.py]: Will revert this.

**Comment by fchollet** [comment]
That's a reasonable change. I would rename the argument `normalize_gradients` to be in line with the Keras 2 API (full words, snake case).

**Comment by raghakot** [comment]
Done.

File: keras/optimizers.py [hunk 2]
```diff
@@ -69,6 +70,9 @@ def get_updates(self, params, constraints, loss):
 
     def get_gradients(self, loss, params):
         grads = K.gradients(loss, params)
+        if hasattr(self, 'normalize_gradients') and self.normalize_gradients:
+            norm = K.sqrt(sum([K.sum(K.square(g)) for g in grads])) + K.epsilon()
```
Comment (fchollet) [inline keras/optimizers.py]: Rather, we should use the `l2_normalize` backend function for better performance. Note that it requires specifying the axis or axes. I believe the correct approach in general is to normalize along all axes.

File: tests/keras/optimizers_test.py [hunk 3]
```diff
@@ -78,6 +78,11 @@ def test_nadam():
     _test_optimizer(optimizers.Nadam())
 
 
+def test_norm_grads():
+    sgd = optimizers.SGD(lr=0.01, momentum=0.9, normalize_gradients=True)
+    _test_optimizer(sgd)
```
Comment (fchollet) [inline tests/keras/optimizers_test.py]: I think we should add a correctness test: a test that checks the updated variables have norm one. We are not doing this for the value clipping test, but that's an oversight.

File: keras/optimizers.py (hunk 2 above)
Comment (fchollet) [inline keras/optimizers.py]: Wondering whether we should allow arbitrary normalization functions (`normalization_fn`) with a built-in L2 option... what do you think?

Comment (raghakot) [inline keras/optimizers.py]: In that case, I would prefer to have gradient_modifier_fn which can allow all sorts of things including clipping etc. There might be use cases where negative gradients are dropped etc. I think it's a nice option for research purposes.

File: tests/keras/optimizers_test.py (hunk 3 above)
Comment (raghakot) [inline tests/keras/optimizers_test.py]: Ah yes. I will fix the other test as well.

File: keras/optimizers.py (hunk 2 above)
Comment (raghakot) [inline keras/optimizers.py]: Btw, here is another use-case: https://arxiv.org/pdf/1707.04199.pdf. Here, the gradients are exponentiated while keeping output activation linear instead of softmax. With `gradient_modfifier_fn` this is easy to handle as well.

Comment (fchollet) [inline keras/optimizers.py]: Let's do `gradient_modifier` then (it's `modifier` so no need for `_fn`, the same way `activation` or `normalizer` and so on don't have the `_fn` suffix). Please add a test case for the use of `lambda x: l2_normalize(x)`. 

Also, feel free to add a basic MLP example script that compares no normalization, L2 normalization, and possibly a few other transformations? Entirely up to you.

Comment (raghakot) [inline keras/optimizers.py]: Sure. I was curious to experiment with some of those changes myself. Might as well test it on simple example.

**Comment by raghakot** [comment]
Looking for some initial feedback since there are a bunch of changes.
I will add tests and doc strings if these look ok.

File: keras/gradient_modifiers.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        if isinstance(then_expression, tf.Tensor):
+            g.set_shape(g_shape)
+        elif isinstance(then_expression, tf.IndexedSlices):
+            g._dense_shape = g_shape
+    else:
+        g = K.switch(K.greater_equal(n, c), g * c / n, g)
+    return g
+
+
+def _norm(gradients, ord):
```
Comment (raghakot) [inline keras/gradient_modifiers.py]: L2 normalize doesn't work with a list of tensors. Hence, the custom implementation.

File: tests/keras/optimizers_test.py [hunk 5]
```diff
@@ -78,14 +78,19 @@ def test_nadam():
...
 
-def test_clipnorm():
-    sgd = optimizers.SGD(lr=0.01, momentum=0.9, clipnorm=0.5)
-    _test_optimizer(sgd)
-
-
-def test_clipvalue():
-    sgd = optimizers.SGD(lr=0.01, momentum=0.9, clipvalue=0.5)
-    _test_optimizer(sgd)
+# def test_norm_grads():
```
Comment (raghakot) [inline tests/keras/optimizers_test.py]: Will update tests after some initial feedback

File: docs/templates/optimizers.md [hunk 6]
```diff
@@ -26,7 +26,16 @@ model.compile(loss='mean_squared_error', optimizer='sgd')
 
 ## Parameters common to all Keras optimizers
 
-The parameters `clipnorm` and `clipvalue` can be used with all optimizers to control gradient clipping:
+The parameters `normalize_gradients`, `clipnorm`, and `clipvalue` can be used with all optimizers to modify gradient.
```
Comment (raghakot) [inline docs/templates/optimizers.md]: Will update this after feedback.

File: keras/optimizers.py [hunk 7]
```diff
@@ -48,19 +20,21 @@ class Optimizer(object):
...
-            when their L2 norm exceeds this value.
-        clipvalue: float >= 0. Gradients will be clipped
-            when their absolute value exceeds this value.
+        gradient_modifier: The gradient modifier to use
+            (see [gradient_modifiers](gradient_modifiers.md)).
     """
 
     def __init__(self, **kwargs):
-        allowed_kwargs = {'clipnorm', 'clipvalue'}
+        allowed_kwargs = {'gradient_modifier'}
```
Comment (raghakot) [inline keras/optimizers.py]: Also wondering if we want to maintain backwards compatibility here. Its easily done but at the cost of making the code unsightly.

Comment (ahundt) [inline keras/optimizers.py]: yes I think that's worthwhile

Comment (Dref360) [inline keras/optimizers.py]: We could deprecate it

File: docs/templates/optimizers.md [hunk 8]
```diff
@@ -26,7 +26,16 @@ model.compile(loss='mean_squared_error', optimizer='sgd')
...
+`normalize_gradients` can be used to improve convergence (see [paper](https://arxiv.org/pdf/1707.04822.pdf) for details) 
+while `clipnorm` and `clipvalue` can be used to manage gradient explosion.
+
+```python
+from keras import optimizers
+
+# Gradients will be normalized to have l2 norm of 1.
+sgd = optimizers.SGD(lr=0.01, norm_grads=True)
```
Comment (Dref360) [inline docs/templates/optimizers.md]: This should be `gradient_modifier` instead of `norm_grads` no?

**Comment by fchollet** [comment]
I recommend:
- Keep backwards compatibility (in particular wrt `clipvalue` and `clipnorm`)
- Do not introduce a new module. I don't think we need built-in modifiers at all.

So this PR should really be:
- A new optimizer kwarg, `gradient_modifier`
- Some check that it's callable (and appropriate error message if not)
- Applying that function in the optimizer
- Serialize it by name in `get_config` (i.e. it will require the custom objects API to load compiled saved models that used a custom gradient modifier).

**Comment by raghakot** [comment]
Should we provide l2_normalize function? or let users implement it themselves? The function can live in optimizers module.

**Comment by fchollet** [comment]
There is one in the backend. Users can use it to specify their own normalization modifier.

**Comment by raghakot** [comment]
That l2_normalize does not work with a list of tensors (which is the case with grads). It normalizes a single tensor.

**Comment by fchollet** [comment]
The `gradient_modifier` function should be applied to each gradient tensor independently (same as weight constraints).

**Comment by raghakot** [comment]
Hmm. The paper (https://arxiv.org/pdf/1707.04822.pdf) doesn't mention it very clearly, but wouldn't we want to normalize across all gradient tensors to preserve relative magnitudes? During backprop, the overall gradient gets distributed and normalizing them independently seems incorrect. Correct me if I am wrong.

**Comment by fchollet** [comment]
I don't know, please check the literature and determine what is the best practice in this case.

**Comment by raghakot** [comment]
Tensorflow docs show that the correct way to `clipnorm` is to use global norm
- https://www.tensorflow.org/api_docs/python/tf/clip_by_global_norm
- Paper ref: https://arxiv.org/pdf/1211.5063.pdf

If `clipnorm` uses global l2 norm then normalize gradients should do the same.

The reason I provided a module is because there are atleast two common use-cases at the moment.
- Injecting gradient noise (lots of recent papers about that).
- Normalizing gradients.

This is another use-case but might not be very common.
- Modified gradients to speed up convergence (ex: https://arxiv.org/pdf/1707.04199.pdf).

We could avoid making a new module, but I think we should provide inbuilt functions for these use-cases given that it's easy to make mistakes (global norm vs local norm). 

The thought behind custom module is so that state information can be maintained within classes. For example, with gradient noise, users might want to configure the mean and variance of the Gaussian noise.

**Comment by fchollet** [comment]
We can provide an example script to clarify how to use gradient
normalization. But adding an entire new modules has far-reaching
implications, and we won't do that now. In the future, if enough people
request this feature, we may add built-in modifiers.

For now, we will just add support for the new keyword argument.

**Comment by raghakot** [comment]
Ok. I will add an example script showing the usage of gradient noise and normalization.

**Comment by fchollet** [comment]
What's the status of this PR?

**Comment by fchollet** [comment]
Sure, no rush! Just wanted to know whether the PR was still active.

**Comment by redouanelg** [comment]
Hi, any news for this PR?
